enhancements/network/user-defined-network-segmentation.md

```diff
@@ -413,6 +669,220 @@ guarantee uniqueness and eliminates the ability to falsify access.
 After creating the CRD, check the status to ensure OVN-Kubernetes has accepted this network to serve the namespaces
 selected. Now tenants may go ahead and be provisioned their namespace.
 
+#### Controller
+
+New controller should be introduced, it should manage the new CRDs lifecycle.
+
+The controller should utilize the CNI plugin API and create OVN networks using `NetworkAttachmentDefinition` using the 
+`ovn-k8s-overlay-cni` CNI plugin according to desired spec.
+
+The controller should watch NAD object in order to reconcile them and reflect the network state in the CRDs status.
+
+The request corresponding NADs should be created with a finalizer, so that upon deletion OVN-Kubernetes can validate that 
+there are no pods still using this network.
+If there are pods still attached to this network, the network will not be removed.
+
+The CRD spec should be validated before creating the NAD, see the [validation](#validations) section for more details.
+
+The controller should create the requested NAD with:
+1. finalizer, enable the controller release resources before NAD is deleted, ensure no pod is connected to the network.
+2. owner-reference referring to the request CR object.
+   Using the owner-reference mechanism should prevent deletion of the NAD before the corresponding CRD instance is deleted.
+   In addition, the owner-reference make teardown seamless, when the request CR object instance is deleted,
+   the cluster garbage collected will dispose the corresponding NAD when all finalizers are gone.
+
+In a scenario a NAD already exist at the target namespace, the controller should check if the existing NAD is managed 
+by the controller as follows:
+1. Check the owner reference match the request CR UID.
+2. OVN-K user-defined-network finalizer exist.
+3. NAD spec correspond to desired spec.
+
+In case one of the previous checks fails, the controller should reconcile the request and 
+this state in the CR status network is not ready.
+
+In scenario of cluster-scope CRD request, the controller should do best effort to create the desired NAD in each
+specified namespace.
+In case on or more NAD creation fails (e.g.: due to namespace not exist), the controller should continue to the 
+next namespace.
+When finished, reflect in the status all namespaces where NAD creation failed to create.
+
+##### General flow
+1. On namespaced-scope CRD creation:
+   - Validates the CRD spec. 
+   - Generate NAD manifest from the CRD spec.
+   - Check that the desired NAD does not already exist. If not, create the NAD and return.
+   - Otherwise, verify the existing NAD correspond to desired spec, if so return.
```

Comment:
so if a NAD already exists  and matches we find that acceptable. One caveat here...since we are not setting owner reference on this NAD to the CR, then on CR delete we wont automatically clean up this already existing NAD right? Just want to make sure I understand correctly.

Reply:
I think in this case we agreed to not touch NADs without our owner reference and report error even if its spec is exactly what we wanted.

Reply:
@npinaeva Github is weird it shows my comment is posted before yours, please have a look